design-proposals/klip-31-metastore-backups.md

```diff
@@ -0,0 +1,122 @@
+nodes in the KSQL cluster writing to the same remote location. The location is a directory where KSQL will write one or more files.
+
+The backup file name will be formed using the command_topic name and the date of the initial backup (i.e. _confluent-ksql-default__command_topic_1593540976.bak).
+
+The messages will be serialized using the command_topic JSON serializer. The backup file content will look pretty similar to what is seen in the command_topic messages. It will have one
+command per line. Each line will start with the same command_topic row timestamp to allow users identify the backup file with the most recent entries (in case of multiple nodes).
+
+i.e. (backup: _confluent-ksql-default__command_topic_1593545289.bak)
+```
+1593545289 {"statement":"CREATE STREAM TEST1(ID INT) WITH(KAFKA_TOPIC='test1', VALUE_FORMAT='JSON');","streamsProperties":{}, ... }
+1593545382 {"statement":"CREATE STREAM TEST2(ID INT) WITH(KAFKA_TOPIC='test2', VALUE_FORMAT='JSON');","streamsProperties":{}, ... }
+1593545529 {"statement":"DROP STREAM TEST1;","streamsProperties":{}, ... }
+```
+
+This format will allow users to easily restore the command_topic using the Kafka producer scripts.
+
+i.e.
+```
+$ cut -d' ' -f2- _confluent-ksql-default__command_topic_1593545289.bak | 
+      kafka-console-producer --broker-list localhost:9092 --topic _confluent-ksql-default__command_topic
+```
+
+### Workflow
+
+When KSQL starts, it always reads the command topic from the beginning. If the backup file does not exist yet, then it will create a new file and start appending each command_topic message to it.
+A backup file is not found if it is the first time backups are enabled, or if the KSQL service.id is different from the ones found in the backup filenames.
+
+If a backup file with the same KSQL service.id exists, then KSQL will append only the new command_topic messages to that file.
+
+KSQL will compare each consumed command_topic message against the next line from the backup file. If both messages are different, then it considers the command_topic as a new restored
+topic, so KSQL will create a new backup file, copy all read messages to the file, and continue appending the new messages to it. If there are no more messages read from the backup file (EOF),
+then KSQL considers this file is the current one and will continue appending messages to it.
+
+KSQL will not use offsets to compare against the backup file. It will assume each offset starts from 1, beginning consuming messages from the topic and reading messages from the file.
+
+### Performance
+
+The KSQL metastore is not expect to grow into a big file. Appending each new command to a file is faster. KSQL command messages are not large to cause slowness during backup file writing.
+It will certainly add extra time during KSQL restarts, but considering the average I/O latency of 10ms per write, then 500 commands would take around 5s to create the initial backup. Also, using
```

Comment:
should be even faster given the whole file is probably just one bulk I/O load

Reply:
We should consider storing a compressed version of the file on disk, to save space. Given that it is pure text, even simple gzip compression should result in a good compression ratio.